design-proposals/VM-cloning.md

```diff
@@ -0,0 +1,112 @@
+
+## Definition of Users
+Everyone who creates a VM might create a VM clone.
+
+## User Stories
+* As a user / admin, I want to be able to create a new VM that is based on an existing VM.
+* As a user / admin, I want to create a clone to a VM using virtctl.
+
+## Repos
+Kubevirt/kubevirt
+
+# Design
+Design can be inspired by the way VMIs are being migrated. In order to migrate a VMI, a [VirtualMachineInstanceMigration](https://kubevirt.io/user-guide/operations/live_migration/#initiate-live-migration)
+CRD is posted into the cluster. This CRD is very thin and basically triggers a migration with basic parameters.
+
+For VM cloning a similar approach can be taken: a new CRD would be introduced that would provide some basic parameters like:
+* The VM that needs to be cloned
+* A new name for the VM (could be optional - a new name will be generated)
+* New MAC address (optional: if not provided MAC would be erased)
+* Whether to clone or skip labels / annotations etc.
+
+More parameters could be added in the future, like the new namespace for the VM, and more.
+
+## API Examples
+As said above, API can be similar to [VirtualMachineInstanceMigration](https://kubevirt.io/user-guide/operations/live_migration/#initiate-live-migration).
+
+Here's a sketch for the new CRD:
+```yaml
+kind: VirtualMachineClone
+  spec:
+    sourceVirtualMachineName: my-cool-vm
+    targetVirtualMachineName: my-cool-clone   # Optional - a new name can be generated, e.g. my-cool-vm-jfh54b 
+    newMacAddress: my-new-mac-address         # Optional - new mac address can be generated automatically
+    
+    # All of the fields below are optional and defaults to true
+    cloneAnnotations: true
+    cloneLabels: true
+    cloneDisks: true
+    cloneNetwork: true
+```
+
+In the future, we may want to support configuring new namespace for the cloned VM. However, this requires
+copying config-maps / secrets to another namespace which could be a potential risk.
+
+## Scalability
+I don't see any scalability issues.
+
+## Update/Rollback Compatibility
+New API so should not affect updates / rollbacks.
+
+## Implementation details / challenges (in short)
+Most of the implementation can be fairly simple, but there are a couple of things that better be in mind
+
+* Storage:
+    * Things are pretty easy thanks to CDI's [smart-cloning](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/smart-clone.md).
+With smart cloning we can simply reference the existing DVs to smart-clone them. Smart cloning is supported only for
+offline usage, therefore it is a current limitation for VM cloning.
+    *  Everything that is backed by a container-disk can be copied very easily, that includes CDRoms, ephemeral disks,
+  etc.
+    * Since secrets and config-maps are namespace scoped, if the VM is cloned to the same namespace it's no problem
+  to reference the same secret / config-maps in the new VM. This ia a reason for limiting the clone to be at the same
+      namespace.
```

Comment:
What about [host disks](https://kubevirt.io/user-guide/virtual_machines/disks_and_volumes/#hostdisk)? They are not copied automatically

Reply:
Host disks are problematic in various ways, and really should be retired.

Reply:
Same here - I think it's best adding to the doc that host disks wouldn't be supported for this.

BTW - since host disks should be retired - maybe we can deprecate it all-together (of course over time etc)?

Reply:
Well its my opinion they should be retired. We should start a discussion, but definitely note that hostdisks would not be supported for cloning.